src/chain_parsers/visualsign-ethereum/src/abi_registry.rs

```diff
@@ -0,0 +1,239 @@
+        Arc::get_mut(&mut self.abis)
+            .expect("ABI map should be mutable")
+            .insert(name.to_string(), Arc::new(abi));
+        Ok(())
+    }
+
+    /// Maps a contract address to an ABI name for a specific chain
+    ///
+    /// # Arguments
+    /// * `chain_id` - The blockchain chain ID (e.g., 1 for Ethereum Mainnet)
+    /// * `address` - The contract address
+    /// * `abi_name` - The ABI name (must be previously registered)
+    pub fn map_address(&mut self, chain_id: ChainId, address: Address, abi_name: &str) {
+        Arc::get_mut(&mut self.address_mappings)
+            .expect("Address mappings should be mutable")
+            .insert((chain_id, address), abi_name.to_string());
+    }
+
+    /// Gets the ABI for a specific contract address on a given chain
+    ///
+    /// # Arguments
+    /// * `chain_id` - The blockchain chain ID
+    /// * `address` - The contract address
+    ///
+    /// # Returns
+    /// * `Some(Arc<JsonAbi>)` if address is mapped and ABI is registered
+    /// * `None` if address is not mapped or ABI not found
+    pub fn get_abi_for_address(&self, chain_id: ChainId, address: Address) -> Option<Arc<JsonAbi>> {
+        let abi_name = self.address_mappings.get(&(chain_id, address))?;
+        self.abis.get(abi_name).cloned()
+    }
+
+    /// Gets an ABI by name
+    ///
+    /// # Arguments
+    /// * `name` - The ABI name (as registered with `register_abi`)
+    ///
+    /// # Returns
+    /// * `Some(Arc<JsonAbi>)` if ABI is registered
+    /// * `None` if ABI not found
+    pub fn get_abi(&self, name: &str) -> Option<Arc<JsonAbi>> {
+        self.abis.get(name).cloned()
+    }
+
+    /// Lists all registered ABI names
+    pub fn list_abis(&self) -> Vec<&str> {
+        self.abis.keys().map(|s| s.as_str()).collect()
+    }
+
+    /// Lists all address mappings for a given chain
+    pub fn list_mappings_for_chain(&self, chain_id: ChainId) -> Vec<(Address, &str)> {
+        self.address_mappings
+            .iter()
+            .filter(|((cid, _), _)| *cid == chain_id)
+            .map(|((_, addr), name)| (*addr, name.as_str()))
+            .collect()
+    }
+}
+
+impl Default for AbiRegistry {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const TEST_ABI: &str = r#"[
+        {
+            "type": "function",
+            "name": "transfer",
+            "inputs": [
+                {"name": "to", "type": "address"},
+                {"name": "amount", "type": "uint256"}
+            ],
+            "outputs": [{"name": "", "type": "bool"}],
+            "stateMutability": "nonpayable"
+        }
+    ]"#;
+
+    #[test]
+    fn test_register_and_retrieve_abi() {
+        let mut registry = AbiRegistry::new();
+        registry
+            .register_abi("TestToken", TEST_ABI)
+            .expect("Failed to register ABI");
+
+        let abi = registry.get_abi("TestToken");
+        assert!(abi.is_some());
```

Comment:
```suggestion
assert_eq!(registry.get(&test_address), Some(&TEST_ABI));
```

to ensure the value stored is the expected value instead of just validating it has Something